components/script/dom/htmlelement.rs

```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
+        // document.
+        let fragment = self.rendered_text_fragment(input);
+
+        // Step 2: Replace all with fragment within element.
+        Node::replace_all(Some(fragment.upcast()), self.upcast::<Node>());
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-outertext>
+    fn GetOuterText(&self) -> Fallible<DOMString> {
+        let node = self.upcast::<Node>();
+        let window = window_from_node(node);
+        let element = self.as_element();
+
         // Step 1.
+        let element_not_rendered = !node.is_connected() || !element.has_css_layout_box();
+        if element_not_rendered {
+            return Ok(node.GetTextContent().unwrap());
+        }
+
+        window.layout_reflow(QueryMsg::ElementOuterTextQuery);
+        let text = window
+            .layout()
+            .query_element_outer_text(node.to_trusted_node_address());
+
+        Ok(DOMString::from(text))
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute%3Adom-outertext-2>
+    fn SetOuterText(&self, input: DOMString) -> Fallible<()> {
+        // Step 1: If this's parent is null, then throw a "NoModificationAllowedError" DOMException.
+        let Some(parent) = self.upcast::<Node>().GetParentNode() else {
+            return Err(Error::NoModificationAllowed);
+        };
+
+        let node = self.upcast::<Node>();
         let document = document_from_node(self);
 
-        // Step 2.
-        let fragment = DocumentFragment::new(&document);
+        // Step 2: Let next be this's next sibling.
+        let next = node.GetNextSibling();
 
-        // Step 3. The given value is already named 'input'.
+        // Step 3: Let previous be this's previous sibling.
+        let previous = node.GetPreviousSibling();
 
-        // Step 4.
-        let mut position = input.chars().peekable();
+        // Step 4: Let fragment be the rendered text fragment for the given value given this's node
+        // document.
+        let fragment = self.rendered_text_fragment(input);
 
-        // Step 5.
-        let mut text = String::new();
+        // Step 5: If fragment has no children, then append a new Text node whose data is the empty
+        // string and node document is this's node document to fragment.
+        if fragment.upcast::<Node>().children_count() == 0 {
+            let text_node = Text::new(DOMString::from("".to_owned()), &document);
 
-        // Step 6.
-        while let Some(ch) = position.next() {
-            match ch {
-                '\u{000A}' | '\u{000D}' => {
-                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
-                        // a \r\n pair should only generate one <br>,
-                        // so just skip the \r.
-                        position.next();
-                    }
+            fragment
+                .upcast::<Node>()
+                .AppendChild(text_node.upcast())
+                .unwrap();
```

Comment:
We should either `expect` with the appropriate message if this cannot fail at this point, or return the error(even though the spec is silent on that).